### ISSUE TITLE
CheckstyleValidatorTest have much code redundancy

### ISSUE THREAD

**Issue by HDouss** [teamed/qulice#526]
Many methods in `CheckstyleValidatorTest` use the same code logic. This code logic is already encapsulated in `validateCheckstyle` method and is used by test methods : `acceptsDefaultMethodsWithFinalModifiers`, `acceptsSingleNameAuthor`, `acceptsConstantUsedInMethodAnnotation`, `acceptsConstructorParametersNamedJustLikeFields`.

Other test methods are not using `validateCheckstyle` method and the same code is repeated over and over. These methods are : `catchesCheckstyleViolationsInLicense`, `acceptsInstanceMethodReferences`, `reportsErrorWhenParameterObjectIsNotDocumented`, `passesWindowsEndsOfLineWithoutException`, `testWindowsEndsOfLineWithLinuxSources`.
This makes the test class unusefully long and unmaintainable.

**Comment by krzyk** [comment]
@davvd valid bug

**Comment by davvd** [comment]
> @davvd valid bug

@krzyk tagged this issue with "bug"

**Comment by krzyk** [comment]
@aschworer, @happy-neko, @jrdalpra, @prahladyeri can you work on this?

**Comment by prahladyeri** [comment]
@davvd Though each method specified in `CheckstyleValidatorTest` "look" alike, they actually aren't. I went through the [code](https://github.com/teamed/qulice/blob/c38a811bfec5453dda9f842d9b82f70e08c08f68/qulice-checkstyle/src/test/java/com/qulice/checkstyle/CheckstyleValidatorTest.java) for this class and found that each mentioned method invokes `validateCheckStyle()` a different Validator each time.

For instance, `acceptsDefaultMethodsWithFinalModifiers()` invokes it with `DefaultMethods.java`, the `acceptsSingleNameAuthor` invokes it with `AuthorTag.java`, etc. So, it doesn't appear that all these methods are in fact, redundant. Do you want me to perform a more rigorous scenario testing to analyze the effects of this method?

**Comment by HDouss** [comment]
@prahladyeri `acceptsDefaultMethodsWithFinalModifiers` and `acceptsSingleNameAuthor` are examples of "good" methods that does not need to be modified. The methods to refactor are : `catchesCheckstyleViolationsInLicense`, `acceptsInstanceMethodReferences`, `reportsErrorWhenParameterObjectIsNotDocumented`, `passesWindowsEndsOfLineWithoutException`, `testWindowsEndsOfLineWithLinuxSources`.

**Comment by krzyk** [comment]
@davvd assign @prahladyeri please

**Comment by jrdalpra** [comment]
@krzyk I only can take 2 tasks at a time. Sorry.

**Comment by prahladyeri** [comment]
@HDouss  Here is my analysis regarding this issue:

1. The `testWindowsEndsOfLineWithLinuxSources()` fail when your sources have linux style EOL (`/r`), whereas your license has windows style EOL (`/r/n`). The other test `passesWindowsEndsOfLineWithoutException()`, however, fails when any of your source files have a windows style EOL. So again, very very subtle difference in functionality though very similar in appearance.

2. `acceptsInstanceMethodReferences()` makes sure that CheckStyleValidator accepts instance methods as arguments.

3. `catchesCheckstyleViolationsInLicense()` is a general test that makes sure that CheckStyleValidator catches certain violations by feeding it an example that is in violation(`"public class Foo { }`).

4. `reportsErrorWhenParameterObjectIsNotDocumented` makes sure that CheckStyleValidator assuredly fails when a function definition is missing a parameter documentation in the JavaDoc syntax.

To summarize, all these methods perform different tests though some of them may appear to be similar.

@krzyk Please suggest further action.

**Comment by HDouss** [comment]
@prahladyeri I am aware that the tests are different.The issue is about refactoring the code so that redundant code (already encapsulated in validateCheckstyle) disappears. The implementation of all the 5 mentioned methods could be replaced by calls to `validateCheckstyle` method, with proper arguments, so that each will test what have to be tested.

**Comment by prahladyeri** [comment]
@HDouss Sure. After having a detailed look at other implemented methods, I can see how these methods are redundant. What needs to be done here is that instead of repeating the whole pattern of creating a mock environment, a license file, `content` String for matching, calling `MatcherAssert.assertThat`, etc.  we can just call `validateCheckstyle()` because that basically does this same pattern.

@krzyk  I'll go ahead and start making these changes on the `master` branch. Let me know if there is anything I should know about.

**Comment by davvd** [comment]
> @davvd assign @prahladyeri please 

@krzyk OK @prahladyeri please start, the task is yours

**Comment by davvd** [comment]
@hdouss thanks for reporting! I topped your account for 15 mins, transaction 72552567

**Comment by prahladyeri** [comment]
@HDouss I've refactored the following two of the mentioned methods:

* `reportsErrorWhenParameterObjectIsNotDocumented`
* `acceptsInstanceMethodReferences`

[Here](https://github.com/teamed/qulice/pull/536) is the link to pull request. The other three methods aren't that straight-forward since they follow some different pattern than the generic `validateCheckstyle` (such as a different EOL character and a different license string). I will need more time to properly analyze and fix those.

**Comment by HDouss** [comment]
@prahladyeri PR #536 failed checks. I can agree with you on `catchesCheckstyleViolationsInLicense`, but for the two other tests, the license file is irrelevant to the test and the check could be delegated to `validateCheckstyle`. All you have to do is to create external files with Linux/Windows EOL.

**Comment by prahladyeri** [comment]
@HDouss I've fixed the code issues with #536, its passing now. For `passesWindowsEndsOfLineWithoutException` and `testWindowsEndsOfLineWithLinuxSources`, the license file is irrelevant, I agree. But they check for Windows style EOL (`/r/n`) rather than the one used in `validateCheckstyle`. So, in order to refactor them, I will have to refactor `validateCheckstyle` too, and add an `if` condition to it so as to consider windows `EOL`s for these two methods and linux style for others. Shall I go ahead and do that?

**Comment by HDouss** [comment]
@prahladyeri I don't think y

**Comment by prahladyeri** [comment]
@HDouss Sure enough. I've refactored both `passesWindowsEndsOfLineWithoutException` and `testWindowsEndsOfLineWithLinuxSources` as well as part of pull request [536](https://github.com/teamed/qulice/pull/536) - build successful, presently under code review.

**Comment by HDouss** [comment]
@prahladyeri Looks good ! I will close the issue as soon as PR #536 is accepted and merged.

**Comment by prahladyeri** [comment]
@HDouss PR #536 is merged. Please close if this resolves the issue.

**Comment by HDouss** [comment]
@prahladyeri Thanks !

**Comment by prahladyeri** [comment]
@davvd I'm yet to be paid on this issue. Is there any specific procedure?

**Comment by krzyk** [comment]
@prahladyeri It usually takes a day or two

**Comment by davvd** [comment]
> @davvd I'm yet to be paid for this. Is there any specific procedure?

@prahladyeri just wait a little bit

**Comment by davvd** [comment]
@prahladyeri I just added **45 mins** to your account, many thanks for your contribution (72986616).. 42 hours and 53 mins spent here

m=2573, that's why a bonus here for fast delivery

+45 to your rating, your total score is [+45](http://www.netbout.com/b/36189?open=rating)

**Comment by prahladyeri** [comment]
@davvd Thanks!

**Comment by davvd** [comment]
> @davvd Thanks!

@prahladyeri you are welcome
